app/vlselect/traces/jaeger/jaeger.go

```diff
@@ -0,0 +1,397 @@
+// See:
+// 1. https://www.jaegertracing.io/docs/1.70/architecture/apis/#http-json-internal
+// 2. https://github.com/jaegertracing/jaeger/blob/9a45f522422c548827b2f3897affc8170e4a3d8b/cmd/query/app/http_handler.go#L110
+func RequestHandler(ctx context.Context, w http.ResponseWriter, r *http.Request) bool {
+	httpserver.EnableCORS(w, r)
+	startTime := time.Now()
+	path := r.URL.Path
+	if path == "/select/jaeger/api/services" {
+		jaegerServicesRequests.Inc()
+		processGetServicesRequest(ctx, w, r)
+		jaegerServicesDuration.UpdateDuration(startTime)
+		return true
+	} else if strings.HasPrefix(path, "/select/jaeger/api/services/") && strings.HasSuffix(path, "/operations") {
+		jaegerOperationsRequests.Inc()
+		processGetOperationsRequest(ctx, w, r)
+		jaegerOperationsDuration.UpdateDuration(startTime)
+		return true
+	} else if path == "/select/jaeger/api/traces" {
+		jaegerTracesRequests.Inc()
+		processGetTracesRequest(ctx, w, r)
+		jaegerTracesDuration.UpdateDuration(startTime)
+		return true
+	} else if strings.HasPrefix(path, "/select/jaeger/api/traces/") && len(path) > len("/select/jaeger/api/traces/") {
+		jaegerTraceRequests.Inc()
+		processGetTraceRequest(ctx, w, r)
+		jaegerTraceDuration.UpdateDuration(startTime)
+		return true
+	} else if path == "/select/jaeger/api/dependencies" {
+		jaegerDependenciesRequests.Inc()
+		// todo it require additional component to calculate the dependency graph. not implemented yet.
+		jaegerDependenciesDuration.UpdateDuration(startTime)
+		return true
+	}
+	return false
+}
+
+// processGetServicesRequest handle the Jaeger /api/services API request.
+// https://github.com/jaegertracing/jaeger/blob/9a45f522422c548827b2f3897affc8170e4a3d8b/cmd/query/app/http_handler.go#L146
+func processGetServicesRequest(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	cp, err := query.GetCommonParams(r)
+	if err != nil {
+		httpserver.Errorf(w, r, "incorrect query params: %s", err)
+		return
+	}
+
+	serviceList, err := query.GetServiceNameList(ctx, cp)
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot get services list: %s", err)
+		return
+	}
+
+	// Write results
+	w.Header().Set("Content-Type", "application/json")
+	WriteGetServicesResponse(w, serviceList)
+}
+
+// processGetOperationsRequest handle the Jaeger /api/services/<service_name>/operations API request.
+// https://github.com/jaegertracing/jaeger/blob/9a45f522422c548827b2f3897affc8170e4a3d8b/cmd/query/app/http_handler.go#L158
+func processGetOperationsRequest(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	cp, err := query.GetCommonParams(r)
+	if err != nil {
+		httpserver.Errorf(w, r, "incorrect query params: %s", err)
+		return
+	}
+
+	// extract the `service_name`.
+	// the path must be like `/select/jaeger/api/services/<service_name>/operations`.
+	u := r.URL.Path[len("/select/jaeger/api/services/"):]
+
+	// check for invalid path: /select/jaeger/api/services/operations
```

Comment:
This check is actually covered by:
```
} else if strings.HasPrefix(path, "/select/jaeger/api/services/") && strings.HasSuffix(path, "/operations") {
``` 
in `RequestHandler`. I'm fine adding extra check for it tho.

Reply:
It is not, I've got a panic, see in commit description.